### ISSUE TITLE
vs.code Go forkbombs Mac with very large numbers of Go compilations

### ISSUE THREAD

**Issue by smarterclayton** [Microsoft/vscode-go#252]
On El Capitan with version 0.10.11 and the Go tools installed, "build on save" has recently started attempting to build each individual package in my (very large) project in parallel, resulting in hundreds of Go compilations running at any one time and eventually wedging the machine due to memory / CPU contention.  If I `pgrep 6g | kill` the processes I can recover, but it happens every time something kicks off a change to the go file.  The number of processes is much more than a go build would use normally, so it looks like it's trying to build every package in my go path at once.

I don't recall seeing this until the most recent vscode update, however we have made other changes to our build structure and it's possible this is something we introduced.

Scenario:

1. Golang 1.4.2 from home-brew
2. Using github.com/openshift/origin (very large package and dependency tree)
3. Base directory `~/projects/origin/src/github.com/openshift/origin`, 
4. GOPATH set via workspace settings to `~/projects/origin/src/github.com/openshift/origin/Godeps/_workspace:~/projects/origin`
5. "Build on save" turned on

Is there a way to get traces on which code is starting the compilation for the packages to get addition debug info?

**Comment by ironcladlou** [comment]
Interesting, I haven't hit this. Although, I'm now using 1.5.3 (I've found it's actually fine for origin now that compiler cache works right for `install`).

Debugging the plugin easy. [These instructions](https://github.com/Microsoft/vscode-go#building-and-debugging-the-extension) are good. It's fun to have an interactive debugger again!

**Comment by ironcladlou** [comment]
Build on save/linting is [implemented here](https://github.com/Microsoft/vscode-go/blob/master/src/goCheck.ts).

**Comment by lukehoban** [comment]
There's three things this could be:

1. Build on save
2. Completion lists
3. Hover


I would be a bit surprised  if (1) was the problem, since it just does a pretty vanilla `go build` and only once per save.

There was a change not too long ago to set `autobuild` to true on `gocode`, which means that (2) will try to build dependencies that aren't built yet when you ask for completion lists.  It's possible that something is going wrong that is causing that to not be able to reuse the build cache.

As mentioned in https://github.com/Microsoft/vscode-go/issues/257#issuecomment-199512489, (3) has changed recently to get more CPU intensive and to make a call to `godoc`.  That could in principle be the issue.

Anyway you can isolate which of these three features is causing the slowdown and the process count explosion in your case?

**Comment by ironcladlou** [comment]
I can reproduce this independently now. It's definitely a result of `go build` checks from build on save. The problem is probably related to `go build` combined with large projects with deep dependency graphs. I can reproduce the behavior by just doing the following in @smarterclayton's case:

```bash
# Assuming GOPATH is already set up correctly
cd $GOPATH/src/my/package/subdir
# This is what vscode is ultimately running on check:
go build -o $(mktemp) .
```

If the package being built imports packages with lots of transitive dependencies (which is the case with the cited project), you end up with many, many go compiler processes because `go build` does not use the package cache (unlike `go install`).

This might be a fundamental limitation of implementing check with `go build`- it simply doesn't scale due to lack of a package cache. This is why in my Sublime integration I leaned heavily on `go fmt` for first-stage syntax checking and `go install` for second-stage checks. I'm not sure what the answer here will be yet, but unfortunately it means users of large projects will probably need to disable check-on-save or anything that relies heavily on `go build`.

**Comment by lukehoban** [comment]
Interesting.  I've tried playing around inside the `github.com/openshift/origin` project per @smarterclayton's comment.  Certainly a large project.  But I'm still not able to reproduce this problem.

For example, if I go to `go/src/github.com/openshift/origin/pkg/dockerregistry`, then run `go build` - it does indeed take 10s or so.  However, if I have Code open, and edit files like `client.go` in that folder, the first time I try to get a completion list will take 10s, but after that, it will be quick, as will build-on-save.  After that initial 10s, I can edit in that file (and others) with no performance problems, and with no unexpected `go` processes being spawned.

My understanding is that this is because `gocode` with `autobuild=true` set (which `vscode-go` uses by default), will populate the build cache, and `go build` will use the cache (it just won't populate it with new results).  

If you have a more detailed repro for this, that would be great.  Sounds like it may be somewhat more specific to the build environment you have in this project beyond what I've been able to reproduce.

**Comment by ironcladlou** [comment]
>My understanding is that this is because gocode with autobuild=true set (which vscode-go uses by default), will populate the build cache, and go build will use the cache (it just won't populate it with new results).

So I sat down and did a little more work and see exactly what's going on. I'm using a quad-core 2.5ghz i7 MBP with 16GB RAM.

Try this as a test environment (assumes `godep` is available):

```bash
mkdir /tmp/origin
export GOPATH=/tmp/origin
git clone git@github.com:openshift/origin.git $GOPATH/src/github.com/openshift/origin
cd $GOPATH/src/github.com/openshift/origin
# This takes ~12s every time.
time $(pushd pkg/deploy/controller/deploymentconfig ; GOPATH=$(godep path):$GOPATH go build -o 
$(mktemp) . ; popd)
# This takes ~22s every time.
time $(pushd pkg/deploy/controller/deploymentconfig ; GOPATH=$(godep path):$GOPATH go test -copybinary -o $(mktemp) -c . ; popd)
```

In VSCode, using these settings:

```json
"go.gopath": "/tmp/origin/src/github.com/openshift/origin/Godeps/_workspace:/tmp/origin",
```

If I open `pkg/deploy/controller/deploymentconfig/controller.go`, just add a newline on line 1, and save, the code is formatted and `go build` is invoked. Takes the same time as the shell scenario. If I open `pkg/deploy/controller/deploymentconfig/controller_test.go`, add a newline and save, `go test` is invoked, again taking the same time as the shell scenario.

In neither case are any `pkg` directories generated in any $GOPATH directories, because neither `go build` nor `go test` use the package cache.

Here's where things get worse, but in a potentially actionable way: if you make rapid changes/saves to either file within VScode, the `go build` and `go test` processes stack up and quickly chew up all my CPU time. The cumulative effect is they all take several times longer to complete and the jet engines roar. It's likely this is what I observed and assessed to be the same manifestation @smarterclayton reported (although I can't say for sure it's exactly the same because AFAIK @smarterclayton would also have godef and godoc processes in the mix which I've hacked out of my fork for testing- which could make his problems even more pronounced than even I'm seeing).

*Side note:* I had no idea about the new gocode default for the auto-cache feature; I've not used that flag it in the past because I don't like the idea of compilation occurring during completion operations, but I understand the value of it from a user experience perspective. At a quick glance I'm not seeing a configuration for it (now that I'm aware of it I'll be disabling that in my fork if necessary). Thanks for the heads up on that change.

**Comment by lukehoban** [comment]
I tried all of that, and if I never use a completion list, I do see some slow builds.  But I can't make more than 1-2 stack up on my machine.  

If I do ever ask for a completion list, after that I don't see any perf issues.

The code that sets the autobuild option on gocode is here: https://github.com/Microsoft/vscode-go/blob/eb48f2f17671612bf3ab73a2e91d06aa5ce06f52/src/goSuggest.ts#L135.

Any reason why you don't want cached builds during development and to rely on those for performant completion and error checking?

**Comment by ironcladlou** [comment]
I took a look at how the gocode auto-compiler is implemented, and it *might* work consistently in theory since it's able to specify packages to `install` that it knows exist and have source code to compile. I haven't got my mind around all the implications of having an external daemon doing its own mtime checks (like the go standard tooling is responsible for) and auto-compiling out of band with interactive user build flows though.

I do agree that the "coincidental" installation of packages by the suggestion engine does reduce `go build` times as a byproduct. I'm not how I feel about the concept yet. It seems very indirect as a solution. I could theoretically get the same behavior but more predictably by just doing a `go install` on my own.

As an aside, and let me know if this would be better discussed elsewhere, but there are other problems with using `gocode` I'm wondering if you've considered. As I noted [here](https://github.com/ironcladlou/GoTools#gocode-caveats) when developing GoTools, there are big problems using gocode across multiple tooling environments (which extends to multiple instances of VScode itself working on separate projects). Watching the gocode daemon debug output while using vscode, I can confirm all the same issues I documented: the `lib-path` gocode uses for searching is inherited by the environment in which the daemon is first executed and doesn't change unless the daemon is explicitly reconfigured.

For example, I have two separate projects in their own isolated GOPATH trees, both of which share a common dependency. If I kill gocode, load VSCode with project A and do a completion, gocode's lib-path is now set to the GOPATH for project A. When I load project B and do a completion, gocode returns completion candidates from project A, because the gocode lib-path hasn't been re-set. In GoTools, I work around this by setting lib-path before every completion, with the caveat that this interferes with other tooling which doesn't account for lib-path.

Interestingly, the gocode auto compiler uses the correct GOPATH when rebuilding. I haven't looked into its GOPATH resolution for that functionality yet.

**Comment by sgallagher** [comment]
@lukehoban Is it possible that it's a platform-based difference? I am seeing exactly the same behavior every time while running on Fedora 23 (with VSCode 0.10.8 running atop the embedded Node 4.1.1). I have it configured to use Go 1.5.3 as installed with `gvm install go1.5.3`.

I have the following workspace settings:

```
// Place your settings in this file to overwrite default and user settings.
{
	"go.buildOnSave": false,
	"go.lintOnSave": false,
	"go.vetOnSave": false,
	"go.formatOnSave": false,
	"go.gopath": "/home/sgallagh/go",
	"go.goroot": "/home/sgallagh/.gvm/gos/go1.5.3",
	"go.formatTool": "goreturns",
	"editor.insertSpaces": false
}
```

So this should not be in any way associated with on-save actions, as I've disabled those. What I see is basically any time that I make a modification that might try to display help in the GUI. (For example, if I type a period/dot, I immediately end up in a saturated CPU situation).

It definitely looks like `gocode` is being called and is farming out dozens of concurrent `go install` commands:

```

sgallagh 29425     1  2 10:30 tty2     00:05:53 /home/sgallagh/go/bin/gocode -s -sock unix -addr 127.0.0.1:37373
sgallagh 29650 20455  0 14:55 tty2     00:00:00 /home/sgallagh/go/bin/gocode -f=json autocomplete /home/sgallagh/go/src/github.com/openshift/origin/test/integration/oauth_basicauth_test.go c9609
sgallagh 29780 29425 14 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util
sgallagh 29856 29425 13 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util
sgallagh 29858 29425 15 14:55 tty2     00:00:00 [go] <defunct>
sgallagh 29860 29425 12 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util
sgallagh 29862 29425  8 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util
sgallagh 29872 29425 13 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util
sgallagh 29895 29425 12 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util
sgallagh 29899 29425 10 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util
sgallagh 29904 29425  9 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util
sgallagh 29917 29425 12 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util
sgallagh 29923 29425 10 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util
sgallagh 29934 29425 14 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util
sgallagh 29935 29425  8 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util
sgallagh 29937 29425  9 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util
sgallagh 29940 29425  9 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util
sgallagh 29942 29425  8 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util
sgallagh 29945 29425 13 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util
sgallagh 29949 29425  8 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util
sgallagh 29950 29425  8 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util
sgallagh 29954 29425  6 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util
sgallagh 29955 29425  6 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util
sgallagh 29959 29425  8 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util
sgallagh 29961 29425  7 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util
sgallagh 29965 29425  9 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util
sgallagh 29970 29425  6 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util
sgallagh 29983 29425  5 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util
sgallagh 29993 29425  6 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util
sgallagh 30000 29425  6 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util/server
sgallagh 30015 29425 10 14:55 tty2     00:00:00 go install github.com/openshift/origin/pkg/client/testclient
sgallagh 30020 29425  6 14:55 tty2     00:00:00 go install github.com/openshift/origin/pkg/cmd/dockerregistry
sgallagh 30022 29425  9 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util
sgallagh 30027 29425  6 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util
sgallagh 30040 29425 10 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util/server
sgallagh 30137 29425  7 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util
sgallagh 30247 29425  7 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util/server
sgallagh 30252 29425  6 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util/server
sgallagh 30273 29425  7 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util/server
sgallagh 30329 29425 23 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util/server
sgallagh 30395 29425  6 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util/server
sgallagh 30522 29425  6 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util
sgallagh 30551 29425  2 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util/server
sgallagh 30570 29425  4 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util
sgallagh 30670 29425  4 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util/server
sgallagh 30678 29425  2 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util/server
sgallagh 30698 29425  1 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util/server
sgallagh 30705 29425  0 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util/server
sgallagh 30726 29425  1 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util/server
sgallagh 30739 29425  0 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util/server
sgallagh 30740 29425  0 14:55 tty2     00:00:00 go install github.com/openshift/origin/test/util/server
sgallagh 30747 29425  0 14:55 tty2     00:00:00 [gocode]
```

This takes between five and fifteen seconds to run, during which time my system is largely unusable since nothing else gets sufficient CPU time.

**Comment by lukehoban** [comment]
@sgallagher I tried again to set up an environment to repro this, but haven't yet been able to.  I still see that after the fast slow completion list request, later requests are returned quickly.  In between, gocode fills up the `~/go/src/github.com/openshift/origin/Godeps/_workspace/pkg/` folder with cached build outputs.

A few thoughts:
1. If you have complete steps for a repro - I could try those.
2. Ultimately though, whatever problem there might be here is almost certainly an issue with `gocode`.  It may make sense to open an issue there instead https://github.com/nsf/gocode.
3. `vscode-go` turns on the `autobuild` option for `gocode`.  That may be what's leading to the issue.  As well as figuring out what's going wrong with `gocode` here, we could also add an option to `vscode-go` to turn off `autobuild`.  In that case, you will only get completion if you have done a local go install manually, and will get results only against the most recently built binaries for cross-package dependencies.  I'm not sure that experience will be better in practice for you.

**Comment by sgallagher** [comment]
@lukehoban So, this is very interesting: the problem goes away entirely if I run `gocode -s -debug` manually before starting VSCode. It does as you say: takes a long time once and then runs fast after that.

However, if the gocode server is launched by VSCode instead, I can always repro the above behavior.

Also, I found this gem in https://github.com/nsf/gocode/issues/333 : 
"Hm, "autobuild" was always experimental and personally I've never used it. So I have no idea how to help you. You can take a look at gocode's code yourself and in particular at autobuild related bits." 

Seems like autobuild isn't really "supported" by gocode upstream.

**Comment by sgallagher** [comment]
So, whatever is happening here is definitely related (somehow) to how vscode-go is launching the gocode service. If I manually call `gocode status` or explicitly call the same command-line that is visible in the process table before launching VSCode, there is no performance issue at all. (I also tried launching it with `gocode set autobuild true`).

So I guess manually starting the gocode service ahead of time is a workaround for now. I can't get the issue to reproduce in a way that would allow me to see the debug information from gocode. It really seems only to rear its ugly head when the gocode service is started by vscode-go.

@smarterclayton This workaround is probably interesting to you.

**Comment by ironcladlou** [comment]
Here's what I do for the best performance/most control:

1. Disable gocode's autobuild feature https://github.com/Microsoft/vscode-go/pull/283
2. Disable `go.buildOnSave`
3. Use a task to `go install` as my manual compile step

I think #283 and the existing capabilities (`go.buildOnSave` and the task system) are enough to close this issue. Perhaps @sgallagher and/or @smarterclayton can test.

**Comment by lukehoban** [comment]
Thanks @ironcladlou - I've merged #283, and I agree that that is likely the best we can do in `vscode-go` to address this.  It would be great to figure out the underlying issue in `gocode` if possible, but that is likely a longer term issue.

I'm going to close this now - but if there's any thoughts on additional things we can do to address this, feel free to re-raise.
